libraries/AP_GPS/AP_GPS_GSOF.cpp

```diff
@@ -221,29 +222,19 @@ AP_GPS_GSOF::SwapFloat(uint8_t* src, uint32_t pos)
 uint32_t
 AP_GPS_GSOF::SwapUint32(uint8_t* src, uint32_t pos)
 {
-    union {
-        uint32_t u;
-        char bytes[sizeof(uint32_t)];
-    } uint32u;
-    uint32u.bytes[0] = src[pos + 3];
-    uint32u.bytes[1] = src[pos + 2];
-    uint32u.bytes[2] = src[pos + 1];
-    uint32u.bytes[3] = src[pos + 0];
-
-    return uint32u.u;
+    // Could also try using std::copy_backward
+    uint32_t u;
+    memcpy(&u, &src[pos], sizeof(u));
+    return __bswap_32(u);
 }
 
 uint16_t
 AP_GPS_GSOF::SwapUint16(uint8_t* src, uint32_t pos)
 {
-    union {
-        uint16_t u;
-        char bytes[sizeof(uint16_t)];
-    } uint16u;
-    uint16u.bytes[0] = src[pos + 1];
-    uint16u.bytes[1] = src[pos + 0];
-
-    return uint16u.u;
+    // Could also try using std::copy_backward
+    uint16_t u;
+    memcpy(&u, &src[pos], sizeof(u));
+    return __bswap_16(u);
```

Comment:
ArduPilot is pretty much stuffed on non-le devices.

But we should be using `be16toh` here anyway - which does nothing if you're on a be host.

Reply:
Great, I just read the [manpage](https://linux.die.net/man/3/be16toh) on these functions. I did not understand what they `spare-endian.h` was for till after reading the manpage. Do you mind if a add a link to the manpage in `sparse-endian.h in this PR? 

Even with this PR, the floating point byteswapping is still not portable to other OS's. What approach would you recommend for those? I cannot locate builtins for endianness conversion of floating points. My idea would be to use the `__BYTE_ORDER` macro inside of `SwapDouble`/`SwapFloat` to make it conditional to run the swap.

Reply:
> Great, I just read the [manpage](https://linux.die.net/man/3/be16toh) on these functions. I did not understand what they `spare-endian.h` was for till after reading the manpage. Do you mind if a add a link to the manpage in `sparse-endian.h in this PR?

Or another PR.

Reply:
Done! https://github.com/ArduPilot/ardupilot/pull/23761

Reply:
Switched to `be16toh` in e216efb6ef6836c78eafee3f85af652f578ec39c. Same for the 32bit swap.

Reply:
Seems like a good reference for the bitswapping. The GSOF specification specifies the floating point format. I'll go compare and check they match. 
https://stackoverflow.com/questions/2727287/why-is-floating-point-byte-swapping-different-from-integer-byte-swapping